### ISSUE TITLE
JAX (vs Autograd) convergence and runtime problems?

### ISSUE THREAD

**Issue by joacorapela** [google/jax#5284]
I built a toy optimization script comparing JAX with Autograd in the minimization of a function using scipy.optimize.BFGS. Two issues:

1. with a start point x0 further from the minimum BFGS with JAX gradients failed to converge, but it converged with Autograd gradients.

2. with a start point x1 closer to the minimum BFGS converged with both gradients. However its runtime was 2.2 times slower with JAX gradients, despite using jit.

Why could this be happening?

Using the code below I get the following output:

> JAX converged with x0?: False
> JAX converged with x1?: True
> JAX elapsed time with x1: 1.44 sec
> 
> Autograd converged with x0?: True
> Autograd converged with x1?: True
> Autograd lapsed time with x1: 0.65 sec

```
import time
import scipy.optimize
import jax.numpy as jnp
import autograd.numpy as anp
import jax
import autograd

def rosenbrock(x):
    answer = sum(100.0*(x[1:]-x[:-1]**2.0)**2.0 + (1-x[:-1])**2.0)
    return answer

jEvalFunc = jax.jit(rosenbrock)
jGradFunc= jax.jit(jax.grad(rosenbrock))
aEvalFunc = rosenbrock
aGradFunc = autograd.grad(rosenbrock)

x0 = [76.0, 97.0, 20.0, 120.0, 0.01, 1e4]
x1 = [76.0, 97.0, 20.0, 120.0, 0.01, 1e2]
jx0 = jnp.array(x0)
jx1 = jnp.array(x1)
ax0 = anp.array(x0)
ax1 = anp.array(x1)

jOptimRes_x0 = scipy.optimize.minimize(fun=jEvalFunc, x0=jx0, method='BFGS', jac=jGradFunc)
startTime = time.time()
jOptimRes_x1 = scipy.optimize.minimize(fun=jEvalFunc, x0=jx1, method='BFGS', jac=jGradFunc)
jElapsedTime = time.time()-startTime

aOptimRes_x0 = scipy.optimize.minimize(fun=aEvalFunc, x0=ax0, method='BFGS', jac=aGradFunc)
startTime = time.time()
aOptimRes_x1 = scipy.optimize.minimize(fun=aEvalFunc, x0=ax1, method='BFGS', jac=aGradFunc)
aElapsedTime = time.time()-startTime

print("JAX converged with x0?: {}".format(jOptimRes_x0.success))
print("JAX converged with x1?: {}".format(jOptimRes_x1.success))
print("JAX elapsed time with x1: {:0.2f} sec".format(jElapsedTime))
print("")
print("Autograd converged with x0?: {}".format(aOptimRes_x0.success))
print("Autograd converged with x1?: {}".format(aOptimRes_x1.success))
print("Autograd lapsed time with x1: {:0.2f} sec".format(aElapsedTime))
```

**Comment by jakevdp** [comment]
I'm not certain about the execution times, but the convergence is likely due to the fact that JAX does computations in float32 by default. If you set `jax_enable_x64` to True, both starting points converge (see https://jax.readthedocs.io/en/latest/notebooks/Common_Gotchas_in_JAX.html#Double-(64bit)-precision)

Regarding the timing: I would not expect scipy optimization routines that use JAX to be particularly performant, because every time a gradient is evaluated, the array must be copied from numpy to JAX and back. Numpy to JAX on CPU cannot be guaranteed to be done in a zero-copy manner due to constraints in XLA. I suspect this repeated copying of data is to blame for JAX's slow execution time.

If you want jax-native optimization, you might try `jax.experimental.optimizers`, although they are still experimental.

**Comment by shoyer** [comment]
`jax.scipy.optimize.minimize` with `method='BFGS'` should also work. It will compile the entire computation + optimization together, which should have less overhead than using SciPy optimizers.

**Comment by jakevdp** [comment]
Forgot about that - thanks @shoyer!

**Comment by joacorapela** [comment]
Thanks @jakevdp  and @shoyer  for your quick replies.

1. As pointed by @jakevdp  I fixed the convergence problem using

> jax.config.update("jax_enable_x64", True)

2. jax.scipy.optimize.minimize does not run faster than scipy.optimize with JAX gradients. 
Am I doing anything wrong using jax.scipy.optimize.minimize (code below)?                                                                                                                                                                                                                                                                      

> scipy.optimize and JAX grads converged with x0?: False
> scipy.optimize and JAX grads converged with x1?: True
> scipy.optimize and JAX grads elapsed time with x1: 1.382964 sec
>
> jax.scipy.optimize and JAX grads converged with x0?: True
> jax.scipy.optimize and JAX grads converged with x1?: True
> jax.scipy.optimize and JAX grads elapsed time with x1: 1.383739 sec
>
> scipy.optimize and Autograd grads converged with x0?: False
> scipy.optimize and Autograd gards converged with x1?: True
> scipy.optimize and Autograd grads elapsed time with x1: 0.632463 sec

```
import time
import scipy.optimize
import jax
import jax.numpy as jnp
import jax.scipy.optimize
import jax.config
import autograd.numpy as anp
import autograd

jax.config.update("jax_enable_x64", True)

def rosenbrock(x):
    answer = sum(100.0*(x[1:]-x[:-1]**2.0)**2.0 + (1-x[:-1])**2.0)
    return answer

jEvalFunc = jax.jit(rosenbrock)
jGradFunc= jax.jit(jax.grad(rosenbrock))
aEvalFunc = rosenbrock
aGradFunc = autograd.grad(rosenbrock)

x0 = [76.0, 97.0, 20.0, 120.0, 0.01, 1e4]
x1 = [76.0, 97.0, 20.0, 120.0, 0.01, 1e2]
toleranceChange = 1e-9
maxIter = 10000
jx0 = jnp.array(x0)
jx1 = jnp.array(x1)
ax0 = anp.array(x0)
ax1 = anp.array(x1)

jOptimRes_x0 = scipy.optimize.minimize(fun=jEvalFunc, x0=jx0, method='BFGS', jac=jGradFunc)
startTime = time.time()
jOptimRes_x1 = scipy.optimize.minimize(fun=jEvalFunc, x0=jx1, method='BFGS', jac=jGradFunc)
jElapsedTime = time.time()-startTime

j2OptimRes_x0 = jax.scipy.optimize.minimize(fun=jEvalFunc, x0=jx0, method='BFGS')
startTime = time.time()
j2OptimRes_x1 = jax.scipy.optimize.minimize(fun=jEvalFunc, x0=jx1, method='BFGS')
j2ElapsedTime = time.time()-startTime

aOptimRes_x0 = scipy.optimize.minimize(fun=aEvalFunc, x0=ax0, method='BFGS', jac=aGradFunc)
startTime = time.time()
aOptimRes_x1 = scipy.optimize.minimize(fun=aEvalFunc, x0=ax1, method='BFGS', jac=aGradFunc)
aElapsedTime = time.time()-startTime

print("scipy.optimize and JAX grads converged with x0?: {}".format(jOptimRes_x0.fun<1e-6))
print("scipy.optimize and JAX grads converged with x1?: {}".format(jOptimRes_x1.fun<1e-6))
print("scipy.optimize and JAX grads elapsed time with x1: {:f} sec".format(jElapsedTime))
print("")
print("jax.scipy.optimize and JAX grads converged with x0?: {}".format(j2OptimRes_x0.fun<1e-6))
print("jax.scipy.optimize and JAX grads converged with x1?: {}".format(j2OptimRes_x1.fun<1e-6))
print("jax.scipy.optimize and JAX grads elapsed time with x1: {:f} sec".format(j2ElapsedTime))
print("")
print("scipy.optimize and Autograd grads converged with x0?: {}".format(aOptimRes_x0.fun<1e-6))
print("scipy.optimize and Autograd gards converged with x1?: {}".format(aOptimRes_x1.fun<1e-6))
print("scipy.optimize and Autograd grads elapsed time with x1: {:f} sec".format(aElapsedTime))

**Comment by joacorapela** [comment]
@jakevdp I am looking forward to start using JAX in my code, but I cannot do it unless I have a decent runtime performance in LBFGS optimizations. Could you please let me know if this could be achieved in the near future. Thanks.

**Comment by jakevdp** [comment]
One issue is that you're calling Python `sum` when it's better to call `jnp.sum`. With this change, I run your benchmarks & see that the JAX method is about 20% faster than scipy, but still about 2x slower than autograd. There may be room for further optimizations in JAX's BFGS method, but I'm not very familiar with that part of the code.

**Comment by joacorapela** [comment]
Thanks @jakevdp your tip of using jnp.sum. In my laptop jnp.sum makes jax.scipy.optimize 9.4% faster ((runtime_slowest-rutime_faster)/runtime_slowest) than scipy.optimize, both methods using JAX gradients. 

Hower, scipy.optimize using Autograd gradients is 66.2% faster than jax.scipy.optimize using JAX gradients. I also found convergence problems in jax.scipy.optimize, compared to scipy.optimize (#5292). So I am concluding that at the moment, for my application (that relies heavily on BFGS optimizations and is not using GPUs) it is better to use scipy.optimize with Autograd gradients. Am I missing anything?

Thanks again, Joaquin

**Comment by cgarciae** [comment]
@joacorapela 

I am probably doing something wrong, but if you jit minimize its absurdly fast. 

```python
import time
import scipy.optimize
import jax
import jax.numpy as jnp
import jax.scipy.optimize
import jax.config
import autograd.numpy as anp
import autograd

jax.config.update("jax_enable_x64", True)

def rosenbrock(x):
    answer = sum(100.0*(x[1:]-x[:-1]**2.0)**2.0 + (1-x[:-1])**2.0)
    return answer

jEvalFunc = jax.jit(rosenbrock)
jGradFunc= jax.jit(jax.grad(rosenbrock))
aEvalFunc = rosenbrock
aGradFunc = autograd.grad(rosenbrock)

x0 = [76.0, 97.0, 20.0, 120.0, 0.01, 1e4]
x1 = [76.0, 97.0, 20.0, 120.0, 0.01, 1e2]
toleranceChange = 1e-9
maxIter = 10000
jx0 = jnp.array(x0)
jx1 = jnp.array(x1)
ax0 = anp.array(x0)
ax1 = anp.array(x1)

@jax.jit
def minimize_jit(x0):
    out = jax.scipy.optimize.minimize(fun=rosenbrock, x0=x0, method='BFGS')
    out = out._replace(message=None)
    return out

jOptimRes_x0 = scipy.optimize.minimize(fun=jEvalFunc, x0=jx0, method='BFGS', jac=jGradFunc)
startTime = time.time()
jOptimRes_x1 = scipy.optimize.minimize(fun=jEvalFunc, x0=jx1, method='BFGS', jac=jGradFunc)
jElapsedTime = time.time()-startTime

j2OptimRes_x0 = minimize_jit(jx0)
startTime = time.time()
j2OptimRes_x1 = minimize_jit(jx1)
j2OptimRes_x1.x.block_until_ready()
j2ElapsedTime = time.time()-startTime

aOptimRes_x0 = scipy.optimize.minimize(fun=aEvalFunc, x0=ax0, method='BFGS', jac=aGradFunc)
startTime = time.time()
aOptimRes_x1 = scipy.optimize.minimize(fun=aEvalFunc, x0=ax1, method='BFGS', jac=aGradFunc)
aElapsedTime = time.time()-startTime

print("scipy.optimize and JAX grads converged with x0?: {}".format(jOptimRes_x0.fun<1e-6))
print("scipy.optimize and JAX grads converged with x1?: {}".format(jOptimRes_x1.fun<1e-6))
print("scipy.optimize and JAX grads elapsed time with x1: {:f} sec".format(jElapsedTime))
print("")
print("jax.scipy.optimize and JAX grads converged with x0?: {}".format(j2OptimRes_x0.fun<1e-6))
print("jax.scipy.optimize and JAX grads converged with x1?: {}".format(j2OptimRes_x1.fun<1e-6))
print("jax.scipy.optimize and JAX grads elapsed time with x1: {:f} sec".format(j2ElapsedTime))
print("")
print("scipy.optimize and Autograd grads converged with x0?: {}".format(aOptimRes_x0.fun<1e-6))
print("scipy.optimize and Autograd gards converged with x1?: {}".format(aOptimRes_x1.fun<1e-6))
print("scipy.optimize and Autograd grads elapsed time with x1: {:f} sec".format(aElapsedTime))
```
Output
```
scipy.optimize and JAX grads converged with x0?: False
scipy.optimize and JAX grads converged with x1?: True
scipy.optimize and JAX grads elapsed time with x1: 1.689257 sec

jax.scipy.optimize and JAX grads converged with x0?: True
jax.scipy.optimize and JAX grads converged with x1?: True
jax.scipy.optimize and JAX grads elapsed time with x1: 0.000679 sec

scipy.optimize and Autograd grads converged with x0?: False
scipy.optimize and Autograd gards converged with x1?: True
scipy.optimize and Autograd grads elapsed time with x1: 0.484108
```

**Comment by joacorapela** [comment]
Many thanks @cgarciae
jit compiling jax.scipy.optimize.minimize yield huge runtime performance improvements.
However, the convergence problem I reported in (#5292) are still present in jax. Please see below.

Thanks again, Joaquin

```python
import jax
import jax.numpy as jnp                                                                                                                                                                                                                                                                                                                                                                                                                   
import jax.scipy.optimize
import jax.config
import autograd.numpy as anp
import autograd

jax.config.update("jax_enable_x64", True)

def zakharov_jaxNumpy(x):
    def zakharovFromIndices(x, ii):
        sum1 = jnp.sum(jnp.power(x, 2))
        sum2 = jnp.sum(0.5*ii*x)
        answer = sum1+sum2**2+sum2**4
        return answer

    ii = jnp.arange(1, len(x)+1, step=1)
    answer = zakharovFromIndices(x=x, ii=ii)
    return answer

def zakharov_autogradNumpy(x):
    def zakharovFromIndices(x, ii):
        sum1 = (x**2).sum()
        sum2 = (0.5*ii*x).sum()
        answer = sum1+sum2**2+sum2**4
        return answer

    ii = anp.arange(1, len(x)+1, step=1)
    answer = zakharovFromIndices(x=x, ii=ii)
    return answer

jEvalFunc = jax.jit(zakharov_jaxNumpy)
jGradFunc= jax.jit(jax.grad(zakharov_jaxNumpy))
aEvalFunc = zakharov_autogradNumpy
aGradFunc = autograd.grad(zakharov_autogradNumpy)

x0 = [76.0, 97.0, 20.0, 120.0, 0.01, 1e4]
x1 = [76.0, 97.0, 20.0, 120.0, 0.01, 1e2]
toleranceChange = 1e-9
maxIter = 10000
jx0 = jnp.array(x0)
jx1 = jnp.array(x1)
ax0 = anp.array(x0)
ax1 = anp.array(x1)

@jax.jit
def minimize_jit(x0):
    out = jax.scipy.optimize.minimize(fun=zakharov_jaxNumpy, x0=x0, method='BFGS')
    out = out._replace(message=None)
    return out

jOptimRes_x0 = scipy.optimize.minimize(fun=jEvalFunc, x0=jx0, method='BFGS', jac=jGradFunc)
startTime = time.time()
jOptimRes_x1 = scipy.optimize.minimize(fun=jEvalFunc, x0=jx1, method='BFGS', jac=jGradFunc)
jElapsedTime = time.time()-startTime

j2OptimRes_x0 = minimize_jit(jx0)
startTime = time.time()
j2OptimRes_x1 = minimize_jit(jx1)
j2OptimRes_x1.x.block_until_ready()
j2ElapsedTime = time.time()-startTime

aOptimRes_x0 = scipy.optimize.minimize(fun=aEvalFunc, x0=ax0, method='BFGS', jac=aGradFunc)
startTime = time.time()
aOptimRes_x1 = scipy.optimize.minimize(fun=aEvalFunc, x0=ax1, method='BFGS', jac=aGradFunc)
aElapsedTime = time.time()-startTime

print("scipy.optimize and JAX grads converged with x0?: {}".format(jOptimRes_x0.fun<1e-6))
print("scipy.optimize and JAX grads converged with x1?: {}".format(jOptimRes_x1.fun<1e-6))
print("scipy.optimize and JAX grads elapsed time with x1: {:f} sec".format(jElapsedTime))
print("")
print("jax.scipy.optimize and JAX grads converged with x0?: {}".format(j2OptimRes_x0.fun<1e-6))
print("jax.scipy.optimize and JAX grads converged with x1?: {}".format(j2OptimRes_x1.fun<1e-6))
print("jax.scipy.optimize and JAX grads elapsed time with x1: {:f} sec".format(j2ElapsedTime))
print("")
print("scipy.optimize and Autograd grads converged with x0?: {}".format(aOptimRes_x0.fun<1e-6))
print("scipy.optimize and Autograd gards converged with x1?: {}".format(aOptimRes_x1.fun<1e-6))
print("scipy.optimize and Autograd grads elapsed time with x1: {:f} sec".format(aElapsedTime))
```

**Comment by cgarciae** [comment]
I ran your version and added some prints with the found solutions:

```
scipy.optimize and JAX grads converged with x0?: True
scipy.optimize and JAX grads converged with x1?: True
scipy.optimize and JAX grads elapsed time with x1: 0.140804 sec
[ 1.07557258e-07 -2.56314298e-09  2.47102737e-07 -1.83395703e-07
 -1.35773457e-07  3.40634401e-08]

jax.scipy.optimize and JAX grads converged with x0?: False
jax.scipy.optimize and JAX grads converged with x1?: True
jax.scipy.optimize and JAX grads elapsed time with x1: 0.000442 sec
[ 1.36015728e-09  2.70359824e-09  1.99681451e-08 -1.56991512e-09
  5.89690825e-10 -1.85915893e-08]

scipy.optimize and Autograd grads converged with x0?: True
scipy.optimize and Autograd gards converged with x1?: True
scipy.optimize and Autograd grads elapsed time with x1: 0.054576 sec
[-7.83446187e-09  1.79433087e-08  3.98365883e-08  3.05088107e-08
  1.50198881e-08 -7.08106598e-08]
```

If the solution is a vector of zeros they are all pretty close 😅, for x1 the minimize_jit solution reports convergence but not for x0 which I guess its the problem?

**Comment by joacorapela** [comment]
Thanks again @cgarciae Yes, for the zakharov function the unique global minimum is at zero (please see [here](https://www.sfu.ca/~ssurjano/zakharov.html)) and the JAX convergence problem is for the starting point x0, which is further away from the global minimum.
